qiskit_experiments/framework/transpilation.py

```diff
@@ -0,0 +1,217 @@
+        physical_qubits: The list of new indices for ``circuit``'s qubit indices.
+        n_qubits: Optional qubit size to use for the output circuit. If
+            ``None``, then the maximum of ``physical_qubits`` will be used.
+
+    Returns:
+        The quantum circuit with new qubit indices
+    """
+    if len(physical_qubits) != circuit.num_qubits:
+        raise QiskitError(
+            f"Circuit to map has {circuit.num_qubits} qubits, but "
+            f"{len(physical_qubits)} physical qubits specified for mapping."
+        )
+
+    # if all(p == r for p, r in zip(physical_qubits, range(circuit.num_qubits))):
+    #     # No mapping necessary
+    #     return circuit
+
+    circ_size = n_qubits if n_qubits is not None else (max(physical_qubits) + 1)
+    p_qregs = QuantumRegister(circ_size)
+    p_circ = QuantumCircuit(
+        p_qregs,
+        *circuit.cregs,
+        name=circuit.name,
+        metadata=circuit.metadata,
+        global_phase=circuit.global_phase,
+    )
+    p_circ.compose(
+        circuit,
+        qubits=physical_qubits,
+        inplace=True,
+        copy=False,
+    )
+    return p_circ
+
+
+def _has_calibration(target: Target, name: str, qubits: tuple[int, ...]) -> bool:
+    """Wrapper to work around bug in Target.has_calibration"""
+    try:
+        has_cal = target.has_calibration(name, qubits)
+    except AttributeError:
+        has_cal = False
+
+    return has_cal
+
+
+def check_transpilation_needed(
+    circuits: Sequence[QuantumCircuit],
+    backend: Backend,
+) -> bool:
+    """Test if circuits are already compatible with backend
+
+    This function checks if circuits are able to be executed on ``backend``
+    without transpilation. It loops through the circuits to check if any gate
+    instructions are not included in the backend's
+    :class:`~qiskit.transpiler.Target`.  The :class:`~qiskit.transpiler.Target`
+    is also checked for custom pulse gate calibrations for circuit's
+    instructions.  If all gates are included in the target and there are no
+    custom calibrations, the function returns ``False`` indicating that
+    transpilation is not needed.
+
+    This function returns ``True`` if the version of ``backend`` is less than
+    2.
+
+    The motivation for this function is that when no transpilation is necessary
+    it is faster to check the circuits in this way than to run
+    :func:`~qiskit.transpile` and have it do nothing.
+
+    Args:
+        circuits: The circuits to prepare for the backend.
+        backend: The backend for which the circuits should be prepared.
+
+    Returns:
+        ``True`` if transpilation is needed. Otherwise, ``False``.
+    """
+    transpilation_needed = False
+
+    if getattr(backend, "version", 0) <= 1:
+        # Fall back to transpilation for BackendV1
+        return True
+
+    target = backend.target
+
+    for circ in circuits:
+        for inst in circ.data:
```

Comment:
Do you have performance profile for something like RB which is ISA but circuit is extremely deep? (RB overwrites the transpile method and probably not a good test case though). The most of experiment instances in our library are with short depth circuit and this double loop should be acceptable considering the transpiler overhead, but I'd like to see some extreme case.

Alternatively we can let experiment class report dependency (e.g. `BaseExperiment.__required_gates__`) and we can test against it, which seems very performant but not really maintainable.

Reply:
Yes, I wish Qiskit maintained something like the equivalent of `count_ops` in `QuantumCircuit` as it is built, so it would not be necessary to iterate through the circuit to find its instructions. I wonder if that has been considered before?

I did not feel too bad about this double loop because I think that `transpile` has to do it as well for every transpiler pass that operates on individual instructions, like `BasisTranslator`, so I don't think this could be slower.

RB is a good example of where `_transpiled_circuits` can still be overridden when necessary for performance.

I don't like `BaseExperiment.__required_gates__` unless there is a convenient way to manage it since most experiments have small circuits like you mentioned. I have also wondered about transpiling on the fly -- instead of doing `circ.x(0)`, doing `circ.compose(self.get("x"))` (needs the backend to already be set), but I don't like the way it breaks from idiomatic way of building circuits.